README.md
# Vue.js Project Timeline

Gantt chart-like project timeline for Vue.js.

![](/docs/screenshot.png?raw=true)

1. [Installation](#installation)
2. [Basic usage](#basic-usage)
3. [API](#api)
   1. [Props](#props)
   2. [Projects Array](#projects-array)
      1. [Available fields](#available-fields)
      2. [Team members](#team-members)
      3. [Example](#example)
4. [Contributing](#contributing)


## Installation

```bash
npm install @mvsde/vue-project-timeline
```

## Basic usage

```vue
<template>
  <project-timeline
    :start-month="startMonth"
    :end-month="endMonth"
    :projects="projects"
  />
</template>

<script>
import ProjectTimeline from '@mvsde/vue-project-timeline'

export default {
  components: {
    ProjectTimeline
  },

  data () {
    return {
      startMonth: new Date('2018-11'),
      endMonth: new Date('2019-02'),
      projects: [
        {
          name: 'Visualize vertical interfaces',
          start: new Date('2018-11-20'),
          end: new Date('2019-01-28')
        }
      ]
    }
  }
}
</script>

<style src="@mvsde/vue-project-timeline/dist/ProjectTimeline.css"></style>
```

More advanced demo code can be found within the [demo project](/demo).


## API

### Props

| Name              | Type    | Required | Default | Description |
| ----------------- | ------- |:--------:| ------- | ----------- |
| `startMonth`      | Date    | ☑        |         | First month displayed in the timeline. The first day of the month for the given date will be automatically calculated. |
| `endMonth`        | Date    | ☑        |         | Last month displayed in the timeline. The last day of the month for the given date will be calculated. |
| `projects`        | Array   | ☑        |         | List of projects. See [Projects array](#projects-array) for a detailed format description. |
| `visibleMonths`   | Integer |          | `12`    | Number of visible months. A scrollbar will be displayed if the timeline has more months than this value. |
| `autoScrollToday` | Boolean |          | `false` | Automatically scroll “today” into view. |
| `locale`          | String  |          | `en-US` | Set the localization with a [locale string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#Locale_identification_and_negotiation). |

### Projects array

#### Available fields

| Name           | Type     | Required | Description |
| -------------- | -------- |:--------:| ----------- |
| `name`         | String   | ☑        | Name of the project. |
| `start`        | Date     | ☑        | Start date of the project. |
| `end`          | Date     | ☑        | End date of the project. |
| `plannedStart` | Date     |          | Planned start date of the project. Displayed as a thin line below the project. |
| `plannedEnd`   | Date     |          | Planned end date of the project. Displayed as a thin line below the project. |
| `color`        | String   |          | Set a specific color for a project by using a valid [CSS color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value). |
| `team`         | Array    |          | Team members of the project. See [Team members](#team-members) for a detailed format description. |
| `onClick`      | Function |          | Click handler for the project name. The project name will be a button if a function is provided. |

#### Team members

| Name           | Type     | Required | Description                |
| -------------- | -------- |:--------:| -------------------------- |
| `name`         | String   | ☑        | Name of the team member.   |
| `avatar`       | String   | ☑        | Avatar of the team member. |

#### Example

The following is an example with all fields populated:

```js
[
  {
    name: 'Visualize vertical interfaces',
    start: new Date('2018-11-20'),
    end: new Date('2019-01-28'),
    plannedStart: new Date('2018-11-10'),
    plannedEnd: new Date('2019-01-13'),
    team: [
      {
        name: 'Nellie Bender',
        avatar: 'https://source.unsplash.com/3402kvtHhOo/32x32'
      },
      {
        name: 'Alex Wilkerson',
        avatar: 'https://source.unsplash.com/das6NrjLoM0/32x32'
      }
    ],
    onClick () {
      console.log('Hello World!')
    }
  }
]
```


## Contributing

```bash
# Start development server
npm run serve

# Run JavaScript linter
npm run lint:js

# Run unit tests
npm run test:unit
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# [0.5.0](https://github.com/mvsde/vue-project-timeline/compare/v0.4.1...v0.5.0) (2019-04-01)


### Features

* Add localization option ([25b3cec](https://github.com/mvsde/vue-project-timeline/commit/25b3cec))
* Make labels sticky ([42eb408](https://github.com/mvsde/vue-project-timeline/commit/42eb408))



## [0.4.1](https://github.com/mvsde/vue-project-timeline/compare/v0.4.0...v0.4.1) (2019-03-14)


### Bug Fixes

* **TodayMarker:** Actually scroll into center ([a17d409](https://github.com/mvsde/vue-project-timeline/commit/a17d409))



# [0.4.0](https://github.com/mvsde/vue-project-timeline/compare/v0.3.2...v0.4.0) (2019-03-14)


### Bug Fixes

* **ProjectTimelineTeam:** Remove avatar double margin ([731f9dd](https://github.com/mvsde/vue-project-timeline/commit/731f9dd))


### Features

* **AutoScroll:** Auto-scroll to today marker ([423c057](https://github.com/mvsde/vue-project-timeline/commit/423c057))
* **TodayMarker:** Mark today with a grid stripe ([d09525d](https://github.com/mvsde/vue-project-timeline/commit/d09525d))



## [0.3.2](https://github.com/mvsde/vue-project-timeline/compare/v0.3.1...v0.3.2) (2019-03-12)


### Bug Fixes

* **ProjectTimeline:** Auto-scale a project's end ([61a530a](https://github.com/mvsde/vue-project-timeline/commit/61a530a))
* **ProjectTimelineLabel:** Final $el fix ([d385966](https://github.com/mvsde/vue-project-timeline/commit/d385966))



## [0.3.1](https://github.com/mvsde/vue-project-timeline/compare/v0.3.0...v0.3.1) (2019-03-11)


### Bug Fixes

* **ProjectTimelineLabel:** $el undefined fix ([ca46bcb](https://github.com/mvsde/vue-project-timeline/commit/ca46bcb))
* **ProjectTimelineLabel:** Fix team member color ([79b5c50](https://github.com/mvsde/vue-project-timeline/commit/79b5c50))



# [0.3.0](https://github.com/mvsde/vue-project-timeline/compare/v0.2.0...v0.3.0) (2019-03-11)


### Bug Fixes

* **ProjectTimelineLabel:** Make sure the label color has enough contrast ([c0ab980](https://github.com/mvsde/vue-project-timeline/commit/c0ab980))
* Only use max-content for vertical grid spacing ([64564bd](https://github.com/mvsde/vue-project-timeline/commit/64564bd))


### Features

* **ProjectTimelineLabel:** Set mix-blend-mode ([e337692](https://github.com/mvsde/vue-project-timeline/commit/e337692))
* **ProjectTimelineTeam:** Conditional avatar icon ([6793d25](https://github.com/mvsde/vue-project-timeline/commit/6793d25))



# [0.2.0](https://github.com/mvsde/vue-project-timeline/compare/v0.1.2...v0.2.0) (2019-03-11)


### Features

* **ProjectTimelineTeam:** Moving v-for to outer container and display member avatar only when one is set. Otherwise fallback to member name ([6b75c45](https://github.com/mvsde/vue-project-timeline/commit/6b75c45))



## [0.1.2](https://github.com/mvsde/vue-project-timeline/compare/v0.1.1...v0.1.2) (2019-03-11)


### Bug Fixes

* Add missing 'e' to package.json style property ([34675c3](https://github.com/mvsde/vue-project-timeline/commit/34675c3))



## [0.1.1](https://github.com/mvsde/vue-project-timeline/compare/v0.1.0...v0.1.1) (2019-03-11)


### Bug Fixes

* Add CSS to readme ([e981027](https://github.com/mvsde/vue-project-timeline/commit/e981027))
* update path to common.js bundle in package.json ([2efdd5a](https://github.com/mvsde/vue-project-timeline/commit/2efdd5a))



# 0.1.0 (2019-02-27)


### Bug Fixes

* Don’t include polyfills in build ([b90908e](https://github.com/mvsde/vue-project-timeline/commit/b90908e))
* Remove old styles ([cf855d7](https://github.com/mvsde/vue-project-timeline/commit/cf855d7))


### Features

* Actually show actual dates ([a791bf3](https://github.com/mvsde/vue-project-timeline/commit/a791bf3))
* Add optional project button ([77e87d1](https://github.com/mvsde/vue-project-timeline/commit/77e87d1))
* Add project team ([5d0d2d6](https://github.com/mvsde/vue-project-timeline/commit/5d0d2d6))
* Add vertical pillars ([b413ba5](https://github.com/mvsde/vue-project-timeline/commit/b413ba5))
* Expose default month number ([d011da0](https://github.com/mvsde/vue-project-timeline/commit/d011da0))


demo/projects.js
const projects = [
  {
    name: 'Visualize vertical interfaces',
    start: new Date('2018-11-20'),
    end: new Date('2019-01-28'),
    plannedStart: new Date('2018-11-10'),
    plannedEnd: new Date('2019-01-13'),
    team: [
      {
        name: 'Nellie Bender',
        avatar: 'https://source.unsplash.com/3402kvtHhOo/32x32'
      },
      {
        name: 'Alex Wilkerson',
        avatar: 'https://source.unsplash.com/das6NrjLoM0/32x32'
      }
    ]
  },
  {
    name: 'Streamline customized models',
    color: 'coral',
    start: new Date('2019-01-30'),
    end: new Date('2019-03-05'),
    plannedStart: new Date('2019-01-01'),
    onClick () {
      console.log('Hello World!')
    }
  },
  {
    name: 'Evolve front-end infrastructures',
    color: 'crimson',
    start: new Date('2019-04-10'),
    end: new Date('2019-06-23'),
    team: [
      {
        name: 'Lord ‘Devious’ Brandon',
        avatar: 'https://source.unsplash.com/uvCxkUuImSw/32x32'
      }
    ]
  }
]

export default projects
